src/advancedHMC_MCMC.jl

```diff
@@ -0,0 +1,318 @@
+    y, st = f.chain(adapt(parameterless_type(ComponentArrays.getdata(θ)), [t]), θ, f.st)
+    ChainRulesCore.@ignore_derivatives f.st = st
+    f.prob.u0 .+ (t .- f.prob.tspan[1]) .* y
+end
+
+# ODE DU/DX
+function NNodederi(phi::LogTargetDensity, t::AbstractVector, θ, autodiff::Bool)
+    if autodiff
+        # returns matrix [derivative returns vector(vector)]
+        hcat(ForwardDiff.derivative.(ti -> phi(ti, θ), t)...)
+    else
+        (phi(t .+ sqrt(eps(eltype(t))), θ) - phi(t, θ)) ./ sqrt(eps(eltype(t)))
+    end
+end
+
+# physloglike over problem timespan
+function physloglikelihood(Tar::LogTargetDensity, θ)
+    p = Tar.prob.p
+    f = Tar.prob.f
+
+    autodiff = Tar.autodiff
+    dt = Tar.physdt
+    t = collect(Float64, Tar.prob.tspan[1]:dt:Tar.prob.tspan[2])
+
+    # # compare derivatives(matrix)
+    out = Tar(t, θ[1:(length(θ) - Tar.extraparams)])
+
+    # # this is a vector{vector{dx,dy}}(handle case single u(float passed))
+    if length(out[:, 1]) == 1
+        physsol = [f(out[:, i][1],
+                     θ[((length(θ) - Tar.extraparams) + 1):length(θ)],
+                     t[i])
+                   for i in 1:length(out[1, :])]
+    else
+        physsol = [f(out[:, i],
+                     θ[((length(θ) - Tar.extraparams) + 1):length(θ)],
+                     t[i])
+                   for i in 1:length(out[1, :])]
+    end
+    physsol = hcat(physsol...)
+
+    # # convert to matrix as nnsol
+    nnsol = NNodederi(Tar, t, θ[1:(length(θ) - Tar.extraparams)], autodiff)
+
+    physlogprob = 0
+    n = length(out[1, :])
+    for i in 1:length(Tar.prob.u0)
+        # can add phystd[i] for u[i]
+        physlogprob += logpdf(MvNormal(nnsol[i, :],
+                                       Diagonal(Tar.phystd[i]^2 .* ones(n))),
+                              physsol[i, :])
+    end
+    return physlogprob
+end
+
+# Standard L2 losses training dataset
+function L2LossData(Tar::LogTargetDensity, θ)
+    # matrix(each row corresponds to vector u's rows)
+    nn = Tar(Tar.dataset[length(Tar.dataset)], θ[1:(length(θ) - Tar.extraparams)])
+
+    L2logprob = 0
+    n = length(nn[1, :])
+    for i in 1:length(Tar.prob.u0)
+        # can add l2std[i] for u[i]
+        # for u[i] ith vector must be added to dataset,nn[1,:] is the dx in lotka_volterra
+        L2logprob += logpdf(MvNormal(nn[i, :], Diagonal(Tar.l2std[i]^2 .* ones(n))),
+                            Tar.dataset[i])
+    end
+    return L2logprob
+end
+
+# priors for NN parameters + ODE constants
+function priorweights(Tar::LogTargetDensity, θ)
+    allparams = Tar.priors
+    invparams = allparams[2:length(allparams)]
+    meaninv = [invparam[1] for invparam in invparams]
+    # nn weights
+    nnwparams = allparams[1]
+    varw = nnwparams[2]^2
+    prisw = nnwparams[1] .* ones(length(θ) - Tar.extraparams)
+
+    if Tar.extraparams > 0
+        # ode parameters
+        invparams = allparams[2:length(allparams)]
+        varinv = [invparam[2]^2 for invparam in invparams]
```

Comment:
Don't need to create this vector I think there's a dispatch for `MvNormal(::Vector, ::Real)`

Reply:
each ODE param might have a different std deviations so ill have to pass a vector, but yeah ive changed this part in other parts of the code.